src/Url/Rebind.hs

```diff
@@ -0,0 +1,129 @@
+{-# LANGUAGE
+    OverloadedStrings
+  , BangPatterns
+  , UnboxedTuples
+  , UnboxedSums
+  , MagicHash
+  , RebindableSyntax
+  , ScopedTypeVariables
+  , LambdaCase
+  , RecordWildCards
+  , NamedFieldPuns
+  , ApplicativeDo
+#-}
+
+-- The parser lives in its own module because it uses RebindableSyntax,
+-- which adversely affects inference and error messages.
+module Url.Rebind
+  ( parserUrl
+  ) where
+
+import Prelude hiding ((>>=),(>>),pure)
+import Data.Bytes.Parser.Rebindable ((>>),(>>=),pure)
+
+import Url.Unsafe (Url(..),ParseError(..))
+import Data.Char (ord)
+import Data.Word (Word8)
+import Data.Bytes.Types (Bytes(..))
+import GHC.Exts (Int(I#),Int#,Word#,(+#),(<#),(-#),orI#,(>=#),(==#),(>#))
+import GHC.Word (Word(W#),Word16(W16#))
+import qualified Data.Bytes as Bytes
+import qualified Data.Bytes.Parser as P
+import qualified Data.Bytes.Parser.Latin as P (skipUntil, char2, decWord16)
+import qualified Data.Bytes.Parser.Unsafe as PU
+import qualified GHC.Exts as Exts
+
+-- | Parser type from @bytesmith@
+-- Note: non-hierarchical Urls (such as relative paths) will not currently parse.
+parserUrl :: Bytes -> P.Parser ParseError s Url
+parserUrl urlSerialization = let !(I# len) = Bytes.length urlSerialization in
+  -- TODO: use skipTrailedByEither later
+  P.measure
+    ( P.skipTrailedBy2 EndOfInput (c2w ':') (c2w '/')
+      `P.orElse`
+      pure False
+    )
+  >>= \(I# i1, !colonFirst) ->
+    ( case colonFirst of
+        False ->
+          P.char2 InvalidAuthority '/' '/' >> pure (i1 -# 1# )
+        True ->
+          -- Jump to position zero is unsound!!!
+          PU.jump 0 >> pure 0#
+    )
+  >>= \urlSchemeEnd -> PU.cursor#
+  >>= \userStart -> P.measure_# (P.skipUntil ':')
+  >>= \i3 -> PU.unconsume (I# i3)
+  >>  P.measure_# (P.skipUntil '@')
+  >>= \i4 -> PU.unconsume (I# i4)
+  >>  P.measure_# (P.skipUntil '/')
+  >>= \i5 -> PU.unconsume (I# i5)
+  >>  ( case (i4 >=# i5) `orI#` (i3 ==# i4) of
+          1# -> PU.jump (I# userStart) >> pure userStart
+          _ -> let jumpi = i4 in
+            case i3 <# i4 of
+              1# -> PU.jump (I# (userStart +# jumpi +# 1# )) >> pure (userStart +# i3)
+              _ -> PU.jump (I# (userStart +# jumpi +# 1# )) >> pure (userStart +# i4)
+      )
+  >>= \urlUsernameEnd -> PU.cursor#
+  >>= \urlHostStart ->
+    ( P.skipTrailedBy2# EndOfInput (c2w ':') (c2w '/')
+      -- Why was this orElse here? When it is commented out,
```

Comment:
all of the test urls happen to have ':' or '/' after the host start, but parsing isn't actually supposed to fail here